### PR TITLE
PFW-1062: Mk3 Toggle and Translations

### PR DESCRIPTION
**Pull request by leptun**
Hi everyone,

In this PR I've implemented a way to make the Toggle menus much more efficient.

As it is now in the MK3 branch, all toggle menu options have a translated counterpart:
```
#MSG_FSENSOR_ON
"Fil. sensor  [on]"
"Fil. senzor [zap]"

#MSG_FSENSOR_NA
"Fil. sensor [N/A]"
"Fil. senzor [N/A]"

#MSG_FSENSOR_OFF
"Fil. sensor [off]"
"Fil. senzor [vyp]"
```
This is very wasteful and with this PR I intend to turn that into:
```
#MSG_FSENSOR
"Fil. sensor"
"Fil. senzor"
```

and common for all toggles:
```
#
"Off"
"Vyp"

#
"On"
"Zap"

#
"N/A"
"\x00"
```

@DRracer @mkbel In theory this will result in significantly reduced flash usage. I've made the necessary changes for this feature to work and changed the "Fans check" message to use this stile of toggle. If you think this is worth developing further, please let me know and I'll start changing the rest of the messages to this. I've also added a "toggle" character at the end. I think it looks nice this way. It can be changed easily to blank like before. One thing I noticed is that some translated toggles don't fit or barely fit. Wouldn't it be better if the toggle section (the one between []) was moved 1 or 2 characters to the right? This way everything fits better, especially with translations.
image

@michalxfanta I've noticed that translations are somewhat neglected, especially french and italian. What do you think of sending an email out like you did for beta testers where you ask for help on existing translations and maybe also new languages?

### CODE REVIEW

**Comment by 3d-gussner** [comment]
Good idea.
Few months ago in https://github.com/prusa3d/Prusa-Firmware/issues/1554 I asked why not all messages are stored in `messages.c/h`.

If Prusa wants to have all messages there I am willing to help. In the MK2 branch nearly all messages where stored here, but it changed in MK3 branch. So I am not sure if it makes sense to spend hours and a PR will not be pulled.

In another Fork we started to add some comments to the LCD layout and define limits of messages, here some examples:
Old
```
//01234567890123456789
//Last print failures
// MMU fails  000
// MMU load fails  000
//
//////////////////////
```
New
```
//|01234567890123456789|
//|Total failures      | c=20
//| MMU fails:      000| c=15
//| MMU load fails: 000| c=15
//| MMU power fails:000| c=15
//////////////////////
```

Old
```
//01234567890123456789
//Total failures
// Power failures  000
// Filam. runouts  000
// Crash  X 000  Y 000
//////////////////////
```
New
```
//|01234567890123456789|
//|Total failures      | c=20
//| Power failures: 000| c=14
//| Filam. runouts: 000| c=14
//| Crash   X:000 Y:000| c=7 for Crash
//////////////////////
```

at this moment 9. This is now just for status screens but it may also apply to menu. If we would add these known `c=` values to `messages.c/h` the `lang-check.py` could find issues way faster and we had less to test and issues like
https://github.com/prusa3d/Prusa-Firmware/issues/1541
https://github.com/prusa3d/Prusa-Firmware/issues/1502
https://github.com/prusa3d/Prusa-Firmware/issues/1469
and others.

@DRracer @mkbel @michalxfanta Should we invest time to improve that?
BTW: I also have a Dutch translation if you are interested, reviewed by two native dutch speaking people plus me.

**Comment by leptun** [comment]
@3d-gussner I’ve tried putting these messages in messages.c/h, but strlen_P(_T()) results in wrong length reported in multi-lang mode. See the last commit I pushed.

**Comment by 3d-gussner** [comment]
> _P(_T

Do you mean that the `lang-check.py` reported wrong length during compiling?

**Comment by leptun** [comment]
@3d-gussner no, I mean it happens just like #1980. [On] is [] and [Off] is [f]

**Comment by leptun** [comment]
I realize now it has nothing to do with Strlen_P. Only that it’s missing the first two characters when compiling multi-lang

**Comment by 3d-gussner** [comment]
@leptun That sucks...isn't then there another issue to be fixed instead to remove MSG from `messages.c/h` every time we get this results? looks to me that MSGs now have a status `they-may-work-or-may-not-we-will-find-out-after-compiling-and-testing-by-looking-on-the-lcd-screen`

**Comment by leptun** [comment]
There might be. Not sure what the issue is for now. I’ll look into it.

**Comment by 3d-gussner** [comment]
> There might be. Not sure what the issue is for now. I’ll look into it.



> I realize now it has nothing to do with Strlen_P. Only that it’s missing the first two characters when compiling multi-lang

So if you compile it in EN_ONLY it works without https://github.com/prusa3d/Prusa-Firmware/pull/2092/commits/7ce2fedac71e8bedd1cc3ad494584cd8f1726b71
If you compile it with multiple-language support it misses the 2 chars
- also in english
- or do they miss only when another languages is chosen?

**Comment by leptun** [comment]
single-lang: it works
multi-lang:en:doesn't work
multi-lang:other:doesn't work

**Comment by leptun** [comment]
@3d-gussner Correction: 
single-lang: it works
multi-lang:en:doesn't work
multi-lang:other: works, but only if it also translates the string and doesn't use the english variant.

**Comment by 3d-gussner** [comment]
@leptun Thanks for the fast answer

> @3d-gussner Correction:
> single-lang: it works
> multi-lang:en:doesn't work
> multi-lang:other: works, but only if it also translates the string and doesn't use the english variant.

@leptun thanks for the fast answer.  How i see it there is an issue with multi-lang **en** part as other languages aren't working when the translation is `"\x00"` but work well if they are translated.

**Comment by leptun** [comment]
@3d-gussner What I don't understand is why lang_get_translation offsets everything by 2. This is where it gets messed up. The offset is only applied to English.

**Comment by 3d-gussner** [comment]
> @3d-gussner What I don't understand is why lang_get_translation offsets everything by 2. This is where it gets messed up. The offset is only applied to English.

you mean this?
```
const char* lang_get_translation(const char* s)
{
	if (lang_selected == 0) return s + 2; //primary language selected, return orig. str.
	if (lang_table == 0) return s + 2; //sec. lang table not found, return orig. str.
	uint16_t ui = pgm_read_word(((uint16_t*)s)); //read string id
	if (ui == 0xffff) return s + 2; //translation not found, return orig. str.
	ui = pgm_read_word(((uint16_t*)(((char*)lang_table + 16 + ui*2)))); //read relative offset
	if (pgm_read_byte(((uint8_t*)((char*)lang_table + ui))) == 0) //read first character
		return s + 2;//zero length string == not translated, return orig. str.
	return (const char*)((char*)lang_table + ui); //return calculated pointer
}

```

**Comment by leptun** [comment]
@3d-gussner Yes. notice the return s + 2

**Comment by 3d-gussner** [comment]
```
#if (LANG_MODE == 0) //primary language only

uint8_t lang_select(uint8_t lang) { return 0; }
```
vs 
`	if (lang_selected == 0) return s + 2; //primary language selected, return orig. str.
`
in multiple language.

**Comment by 3d-gussner** [comment]
just checked https://github.com/prusa3d/Prusa-Firmware/blame/MK3/Firmware/language.c and it hasn't been changed for a year except 5 months ago https://github.com/prusa3d/Prusa-Firmware/commit/5b9e7c4fe7359c6eeb173e232ef19e040056fd69

**Comment by 3d-gussner** [comment]
@leptun as the multi language build isn't that good documented and i needed some time to understand a bit how it work, it may take me some time to figure out why Prusa @XPila @mkbel added the `+2` to `lang_selected` for english when multiple language is used.

**Comment by leptun** [comment]
@3d-gussner I think the +2 was added since ISTR() and _I(), both used in _T(), also add `\xff\xff` at the begining of every string. Aaaaannnndddd.... I just realized I forgot to put ISTR() at the begining of "Off" and "On".

So problem solved. It actually works.

**Comment by 3d-gussner** [comment]
> @3d-gussner I think the +2 was added since ISTR() and _I(), both used in _T(), also add `\xff\xff` at the begining of every string. Aaaaannnndddd.... I just realized I forgot to put ISTR() at the begining of "Off" and "On".
> 
> So problem solved. It actually works.

```
#if (LANG_MODE == 0) //primary language only
#define PROGMEM_I2 __attribute__((section(".progmem0")))
#define PROGMEM_I1 __attribute__((section(".progmem1")))
#define PROGMEM_N1 __attribute__((section(".progmem2")))
#define _I(s) (__extension__({static const char __c[] PROGMEM_I1 = s; &__c[0];}))
#define ISTR(s) s
#define _i(s) _I(s)
#define _T(s) s
```

vs.

```
#else //(LANG_MODE == 0)
// section .loc_sec (originaly .progmem0) will be used for localized translated strings
#define PROGMEM_I2 __attribute__((section(".loc_sec")))
// section .loc_pri (originaly .progmem1) will be used for localized strings in english
#define PROGMEM_I1 __attribute__((section(".loc_pri")))
// section .noloc (originaly progmem2) will be used for not localized strings in english
#define PROGMEM_N1 __attribute__((section(".noloc")))
#define _I(s) (__extension__({static const char __c[] PROGMEM_I1 = "\xff\xff" s; &__c[0];}))
#define ISTR(s) "\xff\xff" s
#define _i(s) lang_get_translation(_I(s))
#define _T(s) lang_get_translation(s)
#endif //(LANG_MODE == 0)
#define _N(s) (__extension__({static const char __c[] PROGMEM_N1 = s; &__c[0];}))
#define _n(s) _N(s)

```

@leptun Thanks, learned something more today. Love the community and sharing the knowledge.

**Comment by 3d-gussner** [comment]
> @3d-gussner no, I mean it happens just like #1980. [On] is [] and [Off] is [f]

As you now found the issue I would like to know what @mkbel means with `as it is preferred macro for internationalized strings` in https://github.com/prusa3d/Prusa-Firmware/pull/1980 and if it makes sense to spend time translating the `messages.c/h` or convert everything to what have been done in https://github.com/prusa3d/Prusa-Firmware/pull/1980?

**Comment by leptun** [comment]
@3d-gussner Aaaand it's still broken. Back to the drawing board.
I guess we'll just have to wait for answers...

**Comment by leptun** [comment]
Now it has two extra character in english and two missing character in other languages

**Comment by 3d-gussner** [comment]
@leptun maybe it is the best to get an answer from Prusa which direction they want to go:
`messages.c/h` vs https://github.com/prusa3d/Prusa-Firmware/pull/1980 approach

The `messages.c/h` has the advantage that all MGSs are stored in one file and translations are easier to do, instead to look through all code files. Another advantage can be to make DE, CZ, IT, ... only versions if needed.

**Comment by 3d-gussner** [comment]
> Now it has two extra character in english and two missing character in other languages

Oh no ...

**Comment by 3d-gussner** [comment]
@leptun I would love to have a simulator so checking translations would be easier.

**Comment by leptun** [comment]
@3d-gussner I got home and tried using MSGs again and this time it worked. On all variants it displayed correctly. This was tested on the MK2.5 FW. Can you test this on the MK3 as well? I want to make sure it also works there.

I found that `lcd_printf_P(PSTR("[%s]"), toggle);` can't be used.
Instead I have to use:
```
lcd_putc('[');
lcd_puts_P(toggle);
lcd_putc(']');
```
Update:
Later I found out that using "[%**S**]" instead of "[%**s**]" (notice the uppercase 'S') is actually the fix to this. This isn't documented anywhere and as far as I know, this is not part of standard c.

**Comment by AndrewLuebke** [comment]
@leptun

 

My understanding is that %S is the same as %ls and is not part of the C99 standard but is included for SUSv2 (Single UNIX Specification v2) compatibility.  So %ls is expecting a wide character (wchar_t) or a wide string (wchar_t*).  Does %ls work in place of %S?  If so that is the “standard C” method.

**Comment by leptun** [comment]
@andrewluebke No, it doesn't work. It breaks exactly like when i was using %s. I think I'll leave it as is for now. I noticed they also used %s in some of the sheets menu formatting. Maybe that's why #1980 was needed? (@3d-gussner)

**Comment by 3d-gussner** [comment]
@leptun Good morning. Great news you found the issue. Just started to compile, test and hopefully confirm that it works in all languages.
Wait for the results.

**Comment by 3d-gussner** [comment]
@leptun 
Here some results:
```
Multi-language firmware will be built



Start to build Prusa Firmware ...
Using variant 1_75mm_MK3S-EINSy10a-E3Dv6full
Build options changed, rebuilding all
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/Marlin_main.cpp: In function 'void process_commands()':
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/Marlin_main.cpp:4706:15: warning: unused variable 'kill_message' [-Wunused-variable]
   const char *kill_message = NULL;
               ^
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/lcd.cpp: In function 'void lcd_putchar(char, FILE*)':
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/lcd.cpp:192:39: warning: 'void lcd_putchar(char, FILE*)' was declared 'extern' and later 'static' [-fpermissive]
 static void lcd_putchar(char c, FILE *)
                                       ^
In file included from <path>/Prusa-Firmware/Prusa-Firmware-build/sketch/lcd.cpp:3:0:
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/lcd.h:14:13: note: previous declaration of 'void lcd_putchar(char, FILE*)'
 extern void lcd_putchar(char c, FILE *stream);
             ^
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/ultralcd.cpp: In function 'void lcdui_print_farm()':<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/ultralcd.cpp:642:6: warning: unused variable 'chars' [-Wunused-variable]
  int chars = lcd_printf_P(_N(" F0  "));
      ^
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/ultralcd.cpp: At top level:
<path>/Prusa-Firmware/Prusa-Firmware-build/sketch/ultralcd.cpp:5470:13: warning: 'lcd_check_gcode_set' defined but not used [-Wunused-function]
 static void lcd_check_gcode_set(void)
             ^
<path>/Prusa-Firmware/PF-build-env-1.0.6/linux-64/portable/packages/PrusaResearchRambo/hardware/avr/1.0.1/cores/rambo/wiring.c:40:22: warning: 'timer0_fract' defined but not used [-Wunused-variable]
 static unsigned char timer0_fract = 0;
                      ^
Sketch uses 242950 bytes (95%) of program storage space. Maximum is 253952 bytes.
Global variables use 6045 bytes of dynamic memory.
```

The warnings aren't related to your change and I am not that concenerned about
- `[-Wunused-variable]` or `[-Wunused-function]`				even it could be avoided by some if statements as it was done in MK2 branch
but
- `was declared 'extern' and later 'static' [-fpermissive]`		should be fixed.



**Now the results on MK3S:**
- Toggle works in English, German, French, Spanish and Italian other languages are bit hard for me to understand. **Your change works.**
- What have to be done:
  - In some languages the translations are switching between the two states what is kind of annoying:
    `"F.Autoladen  [an]"` vs `"F. Autoload [aus]"` one example in German there are more
	`"ChargAuto f. [on]"` vs `"AutoCharg F [off]"` one example in French
	`"Autocar.filam[on]"` vs `"Autocar.fil.[off]"` one example in Italian
   - Maybe use bit more of the screen as the last two aren't used (except the "Fans check    [ON/OFF]" has a strange icon behind)
- Other places your change can be applied:
  - Settings -> HW Setup -> Checks
  - Settings -> Temp. cal-
  - Settings -> RPi port
  - SpoolJoin
  - Cutter
  - Magnets comp.
```

#define SETTINGS_MODE \
do\
{\
    switch(oCheckMode)\
         {\
         case ClCheckMode::_None:\
              MENU_ITEM_FUNCTION_P(_i("Nozzle     [none]"),lcd_check_mode_set);\
              break;\
         case ClCheckMode::_Warn:\
              MENU_ITEM_FUNCTION_P(_i("Nozzle     [warn]"),lcd_check_mode_set);\
              break;\
         case ClCheckMode::_Strict:\
              MENU_ITEM_FUNCTION_P(_i("Nozzle   [strict]"),lcd_check_mode_set);\
              break;\
         default:\
              MENU_ITEM_FUNCTION_P(_i("Nozzle     [none]"),lcd_check_mode_set);\
         }\
}\
while (0)

```

```

#define SETTINGS_MODEL \
do\
{\
    switch(oCheckModel)\
         {\
         case ClCheckModel::_None:\
              MENU_ITEM_FUNCTION_P(_i("Model      [none]"),lcd_check_model_set);\
              break;\
         case ClCheckModel::_Warn:\
              MENU_ITEM_FUNCTION_P(_i("Model      [warn]"),lcd_check_model_set);\
              break;\
         case ClCheckModel::_Strict:\
              MENU_ITEM_FUNCTION_P(_i("Model    [strict]"),lcd_check_model_set);\
              break;\
         default:\
              MENU_ITEM_FUNCTION_P(_i("Model      [none]"),lcd_check_model_set);\
         }\
}\
while (0)

```

```

#define SETTINGS_VERSION \
do\
{\
    switch(oCheckVersion)\
         {\
         case ClCheckVersion::_None:\
              MENU_ITEM_FUNCTION_P(_i("Firmware   [none]"),lcd_check_version_set);\
              break;\
         case ClCheckVersion::_Warn:\
              MENU_ITEM_FUNCTION_P(_i("Firmware   [warn]"),lcd_check_version_set);\
              break;\
         case ClCheckVersion::_Strict:\
              MENU_ITEM_FUNCTION_P(_i("Firmware [strict]"),lcd_check_version_set);\
              break;\
         default:\
              MENU_ITEM_FUNCTION_P(_i("Firmware   [none]"),lcd_check_version_set);\
         }\
}\
while (0)

```

```

#define SETTINGS_GCODE \
do\
{\
    switch(oCheckGcode)\
         {\
         case ClCheckGcode::_None:\
              MENU_ITEM_FUNCTION_P(_i("Gcode      [none]"),lcd_check_gcode_set);\
              break;\
         case ClCheckGcode::_Warn:\
              MENU_ITEM_FUNCTION_P(_i("Gcode      [warn]"),lcd_check_gcode_set);\
              break;\
         case ClCheckGcode::_Strict:\
              MENU_ITEM_FUNCTION_P(_i("Gcode    [strict]"),lcd_check_gcode_set);\
              break;\
         default:\
              MENU_ITEM_FUNCTION_P(_i("Gcode      [none]"),lcd_check_gcode_set);\
         }\
}\
while (0)

```

**Comment by leptun** [comment]
@3d-gussner the `was declared 'extern' and later 'static'` warning has been fixed in my LCD buffer implementation.

The icon after "Fans check" was added by me. It can easily be removed. It is one of the unused custom characters on the lcd that for some reason still exists. Maybe it could be removed to save some more space.

Only the "Fans check" message was changed to use this feature for now. That's why this PR is a WIP. I intend to change all toggles in the firmware to this style, not only those few you mentioned. For now I am waiting for confirmation from @DRracer or @mkbel that this is something they want merged since it's a lot of work changing all the translations by hand.

Also about the changing translations between states: This PR will avoid that entirely since "On" will be "On" everywhere, "Off" will be "Off" and "N/A" will be "N/A". On other languages it is not consistent and you might see for example "On" and the next line be "on". This PR should prevent this kind of issues.
Please do the following: flash the firmware without this PR, change the lang to Italian and French and look at the toggles in settings and switch a few of them. You will see what I am saying about the inconsistency.

**Comment by leptun** [comment]
@3d-gussner We could move the toggle two characters to the right (without using that character), effectively having it aligned to the right edge of the screen.
image
Ignore the second line. It was not changed.

**Comment by 3d-gussner** [comment]
> @3d-gussner the `was declared 'extern' and later 'static'` warning has been fixed in my LCD buffer implementation.
> 

Cool

> The icon after "Fans check" was added by me. It can easily be removed. It is one of the unused custom characters on the lcd that for some reason still exists. Maybe it could be removed to save some more space.
> 

That would be good as non-english translations may take more space. Maybe even move to toggle as far right as possible.

> Only the "Fans check" message was changed to use this feature for now. That's why this PR is a WIP. I intend to change all toggles in the firmware to this style, not only those few you mentioned. For now I am waiting for confirmation from @DRracer or @mkbel that this is something they want merged since it's a lot of work changing all the translations by hand.
> 

Understand it is kind of Proof of Concept ... Great Job BTW

> Also about the changing translations between states: This PR will avoid that entirely since "On" will be "On" everywhere, "Off" will be "Off" and "N/A" will be "N/A". On other languages it is not consistent and you might see for example "On" and the next line be "on". This PR should prevent this kind of issues.
> Please do the following: flash the firmware without this PR, change the lang to Italian and French and look at the toggles in settings and switch a few of them. You will see what I am saying about the inconsistency.

Also think we need separate PRs for translations. I will try to make a PR so it is bit more clear how long the translations can be like i mentioned above. 

```

//|01234567890123456789|
//|Total failures      | c=20
//| Power failures: 000| c=14
//| Filam. runouts: 000| c=14
//| Crash   X:000 Y:000| c=7 for Crash
//////////////////////

```

**Review by 3d-gussner — APPROVED** [review]
Tested on MK3S.
I can confirm it works on multi-language version in English, German, French, Spanish and Italian.
Didn't test Czech or Polish as these are bit too hard for me to understand

**Comment by DRracer** [comment]
Wow, great job @leptun @3d-gussner , do you ever sleep ;) ? 
Making the texts more consistent while lowering FLASH usage is of course welcome. Since @leptun is almost ready with the "vga" buffer, this might be a nice bonus to put into FW 3.9 - work will start in late August.
I was looking into the translations scripts lately - fixing the \n issue. I still don't understand the whole process in detail. Right now translators get a *.po file - basically a gettext format exported from the lang_xx.txt files. However, moving text around between lang_xx.txt, removing unused texts, adding new etc. must be performed by hand, may be that's why the content wasn't updated regularly.

My proposal: let's start with a proper documentation of the current state of the language translation processes, then propose improvements and implement an agreed solution. @leptun @3d-gussner, do you agree?

Btw. printf("%S") reads the string from PROGMEM (%s is from RAM)

**Comment by DRracer** [comment]
> We could move the toggle two characters to the right (without using that character)

@leptun We'd like to keep the last character for "->" and your new "refresh" character (btw. nice improvement ;) ) - the same way as we try to keep the first character for menu cursor.

**Comment by leptun** [comment]
@DRracer I agree it would be best if we started with a good understanding of the translation process.

Good thing to know that “%S” is for progmem. I thought it just needed the correct pointer, but I guess it also needs this.

When I get back I’ll change the formatting to have the toggle one space to the right, the last place being the toggle character when it toggles and the arrow character if it’s [N/A].

**Comment by 3d-gussner** [comment]
> Wow, great job @leptun @3d-gussner , do you ever sleep ;) ?

LOL, lately don't sleep that well so more time on computer.

> Making the texts more consistent while lowering FLASH usage is of course welcome. Since @leptun is almost ready with the "vga" buffer, this might be a nice bonus to put into FW 3.9 - work will start in late August.

I have to look into @leptun "vga" buffer ... :-)

> I was looking into the translations scripts lately - fixing the \n issue. I still don't understand the whole process in detail. 
Right now translators get a *.po file - basically a gettext format exported from the lang_xx.txt files. However, moving text around between lang_xx.txt, removing unused texts, adding new etc. must be performed by hand, may be that's why the content wasn't updated regularly.
> 

I asked few times how the process works, but never got an answer so I started to investigate more. Finally i got some idea how the process works and added Dutch. Here some comments https://github.com/prusa3d/Prusa-Firmware/issues/1150#issuecomment-424276886

> My proposal: let's start with a proper documentation of the current state of the language translation processes, then propose improvements and implement an agreed solution. @leptun @3d-gussner, do you agree?
> 

Totally agree. It would be great if the writer of the `/lang/` scripts could add some information. I have some idea how it works and could begin. Where should we place it, in an Issue, Project or Wiki?

> Btw. printf("%S") reads the string from PROGMEM (%s is from RAM)

Thanks for pointing this out.

As I am living in a Dutch speaking country and like to improve the 3d printing experience especially for **children** and older non-english speaking persons, I started to add Dutch translations to the MK2 branch https://github.com/3d-gussner/Prusa-Firmware/releases/tag/v3.0.10-12 and added later French and Portuguese https://github.com/3d-gussner/Prusa-Firmware/releases/tag/v3.1.0-RC2-HF
We have few kids in our maker space and I see them struggling with tools (computers, 3d printers, applications that are just in english). On the other hand it is quite daunting for people to help with translations if they aren't used to code. So you can count on me and others helped me to translate.

In the MK2 branch it was quite easy as you just had to translate from the `language_en.h` and into the `language_<language code>.h` file.
Now in the MK3 branch it looks like there is no standard:
- Sometimes `messages.c/h` is used
- but more and more changes are plain in the code.

Before i write more maybe it is the best if we open a new **issue**, **project** or **Wiki** in github for that.

**Comment by DRracer** [comment]
@3d-gussner I already have some experience with the lang process, I'll document what I know by the end of this week. I'd like to keep a final file of translations.md in the lang dir or so, that ensures everyone gets the doc when pulling the code. However, we may start with a wiki page, that might allow us a faster cooperation

**Comment by 3d-gussner** [comment]
> @3d-gussner I already have some experience with the lang process, I'll document what I know by the end of this week. I'd like to keep a final file of translations.md in the lang dir or so, that ensures everyone gets the doc when pulling the code. However, we may start with a wiki page, that might allow us a faster cooperation

Like the idea of `translations.md` ... and as soon you create one we can easily add comments and PRs

**Comment by leptun** [comment]
@DRracer @3d-gussner I think I understood how to make a language translation. The question remains: are those .po files so important? They were last updated 6 months ago. I think these were relevant a long time ago when big changes were made with the help of the translators. Or what is more likely is that’s from when the language building process migrated from the perl script in firmware to its own folder with bash scripts.
Regardless of the relevance of po/pot files, I think we could start translating from an existing lang_en_xx.txt file. Before that though it would be helpful if all messages were moved to the messages.c/h files. I was thinking we could “sort” the messages in the order they appear in the translation files and have at the beginning of the messages files the _N region. 

Still waiting an answer about the preferred method though... @mkbel @XPila

**Comment by leptun** [comment]
> @michalxfanta I've noticed that translations are somewhat neglected, especially french and italian. What do you think of sending an email out like you did for beta testers where you ask for help on existing translations and maybe also new languages?

@michalxfanta Ping?
I will also provide Romanian translation once everything is decided.

**Comment by DRracer** [comment]
@leptun the PO files are only an interface to external translator people - it is standard gettext format AFAIK. When we get the PO files back, we must manually copy the translated strings into the lang_en_xx.txt files. Btw. we are still waiting for the translated texts for the FW 3.8 ;)

**Comment by GurliGebis** [comment]
@DRracer would it make sense for the compile proces to extract the text from the po files, and generate the language txt files on the fly?

**Comment by leptun** [comment]
@DRracer @3d-gussner I've updated the format as discussed earlier.
In order to test it I also changed the Filament sensor and autoload toggles to this style and moved some messages to messages.c/h.
Here is a photo with it now:
image

**Comment by leptun** [comment]
@DRracer Any update on the documentation?

> @3d-gussner I already have some experience with the lang process, I'll document what I know by the end of this week. I'd like to keep a final file of translations.md in the lang dir or so, that ensures everyone gets the doc when pulling the code. However, we may start with a wiki page, that might allow us a faster cooperation

**Comment by DRracer** [comment]
@leptun yes, indeed ;) . Still WIP, but we've already managed to decode and understand some of the scripts. The referenced translations.md is in my repo for now, please feel free to comment and post PR's ;) .
https://github.com/DRracer/Prusa-Firmware/tree/translations-doc
Still there is a large code in the firmware that handles the translated messages.

**Comment by leptun** [comment]
@DRracer Great. I'll look into it later. Did you find if it's better using _i("") or _T(MSG)?

**Comment by DRracer** [comment]
@leptun the code parts like _i, _T, MSG*... are still not completely clear to me, let's work on it together. There are multiple development approaches of various age mixed together.

**Comment by DRracer** [comment]
@leptun update... It looks like the MSG_xxxx are an older approach... still have to find out
@leptun @3d-gussner @wavexx I've added you as collaborators in my repo, let's decrypt the language process together ;)

**Comment by leptun** [comment]
@DRracer I've accepted the invitation.
I mean the MSG_xxxx approach seems more intuitive and one that has advantages over _i(). You have to edit the string only in one place in the firmware and then in many places in the lang files.

Would there be any performance advance for using _i() or _T() over the other?

**Comment by DRracer** [comment]
@leptun thanks ;) . I don't see any performance benefit using _i over _T or vice versa. They are just pieces of FLASH memory, that need to be read.

**Comment by leptun** [comment]
@DRracer Is there any penalty if a string from progmem is used in multiple places? Does the compiler optimize it differently depending on the method used?

**Comment by DRracer** [comment]
@leptun you are right, _i is not suited well for multiple usage of the same string. That's one of the benefits of using _T. However, _i is really newer approach and was developed just to allow simpler inplace declaration of strings - i.e. avoiding the string table.

**Comment by leptun** [comment]
@DRracer I suppose what is now left to do is figure out why there are two reserved characters at the beginning of each translated string. (ISTR)

I think _T and_i should both be used depending on their specific case:
- If there is more than one the string is used, then we use _T(MSG_) since it keeps the string in one place instead of being scattered.
- If there is only one appearance of a string, then we use _i(STR).
There are some MSGs that are only used in one place. If the method mentioned above is going to be used, then I suggest we also move those MSGs to _i(STR) counterpart since it's simply not needed. Of course we can just leave them as they are now and just focus on new strings, but keeping everything organized should be a priority in my opinion.

I've started translating the strings to Romanian. I see that on the website there are the languages you can find in the firmware. Since some of the strings point you to chapters in the manual with correct naming, should there also be a translated manual counterpart? I'm asking since that seems very tedious and might be a dealbreaker for adding new languages.

**Comment by 3d-gussner** [comment]
> @leptun update... It looks like the MSG_xxxx are an older approach... still have to find out
> @leptun @3d-gussner @wavexx I've added you as collaborators in my repo, let's decrypt the language process together ;)

Thanks for the invitation... Let's start

**Comment by leptun** [comment]
@DRracer Should I continue changing the toggles according to this? 

> @leptun you are right, _i is not suited well for multiple usage of the same string. That's one of the benefits of using _T. However, _i is really newer approach and was developed just to allow simpler inplace declaration of strings - i.e. avoiding the string table.

Also, my MMU2S has arrived🎉. I've assembled it, connected it to power, connected the data cable to my test bench aaaand was greeted with a big load of nothing. After some intense debugging (two days), I came to the conclusion that I've received a MMU2S board that was blank. The Atmega32u4 on the board had it's default fuse settings and the flash was empty as reported by avrdude. I suggest that you report this incident further since it is a serious issue that might also affect other people who bought the MMU2S recently. Luckily, I had a programmer lying around, so I flashed the bootloader (that was surprisingly hard to find) and set the fuses to the board, uploaded the firmware with PrusaSlicer and BAM, now it's working. Hopefully when the time comes to connect it to a real printer everything works smoothly :).

I've looked through the MMU2S firmware and was disappointed to see how basic it is. Why isn't there a fastio implementation, and more importantly, why isn't there any motion planner like in Prusa-Firmware? Anyways, once I get familiarized with the hardware through usage, I'm thinking of implementing these myself.
@DRracer Who is currently developing the MMU firmware? I see that you and @mkbel are the most recent to have committed to the master branch. Also would you accept PRs also there?

**Comment by DRracer** [comment]
@leptun yes please, continue on this set of improvements. Have you checked the translation documentation lately? Anything still unclear/undocumented?

MMU: OMG, so it happened also to you :( . We are getting reports about the MMU boards coming completely blanked even though the kits are being tested properly. So please - pay close attention to this problem - if it happens again, try to download what's left of the FLASH (if anything).
Anyway, only @mkbel was making PRs for the MMU for a year or more, I'd like to have a deeper insight too - that's the plan ;) .

**Comment by 3d-gussner** [comment]
> Luckily, I had a programmer lying around, so I flashed the bootloader (that was surprisingly hard to find) and set the fuses to the board, uploaded the firmware with PrusaSlicer and BAM, now it's working. Hopefully when the time comes to connect it to a real printer everything works smoothly :).
> 

Why it was hard to find the bootloader, as it should be in the Prusa-Arduino-Boards JSON file?

> I've looked through the MMU2S firmware and was disappointed to see how basic it is. Why isn't there a fastio implementation, and more importantly, why isn't there any motion planner like in Prusa-Firmware? Anyways, once I get familiarized with the hardware through usage, I'm thinking of implementing these myself.

> @DRracer Who is currently developing the MMU firmware? I see that you and @mkbel are the most recent to have committed to the master branch. Also would you accept PRs also there?

TheZeroBeast made some improvements in the past to the MMU2 and MK3 firmware so it was reliable, but his attempts to get it into Prusa repositories never made it. I know it is quite old but checkout his firmwares:
https://github.com/TheZeroBeast/TZB-MMU2S-Firmware/tree/MMU-FSensor
https://github.com/TheZeroBeast/TZB-MMU2S-Firmware/releases
https://github.com/TheZeroBeast/TZB-MK3S-Firmware/tree/MK3-MMU-FSensor
https://github.com/TheZeroBeast/TZB-MK3S-Firmware/releases

He just got back to work on it to get it working with MK3s and MMU2s...I hope Prusa will consider his achievements this time to get implemented to the ROOT :-)

**Comment by leptun** [comment]
@3d-gussner That’s where I got it from, but first I thought it had it’s own repo in the prusa3d user. Started compiling Caterina and lufa...

**Comment by DRracer** [comment]
@3d-gussner thanks for pointing out TheZeroBeast

File: Firmware/ultralcd.cpp [hunk 1]
```diff
@@ -5106,30 +5106,24 @@ static void auto_deplete_switch()
...
-        }
-    }
-    return false;
-}
-
 #define SETTINGS_AUTO_DEPLETE \
 do\
 {\
-    if(settingsAutoDeplete()) return;\
+	if (mmu_enabled)\
```
Comment (DRracer) [inline Firmware/ultralcd.cpp]: These long macros may be hard to understand and debug (even the IDE's may have problems understanding this).
IMO the previous version with the settingsAutoDeplete() function call was easier to maintain.
However, I really like the new construct: MENU_ITEM_TOGGLE_P(_i("SpoolJoin"), _T(MSG_NA), NULL);....

Comment (leptun) [reply Firmware/ultralcd.cpp]: @DRracer I don't know how we could make it more readable. Do you mean the fact that it's all in the define with \ at the end? We could move it's contents to a static void function like before.
The thing is I did it the same on the Cutter.

Comment (DRracer) [reply Firmware/ultralcd.cpp]: The static function like before is the preferred solution, thank you ;)

**Review by DRracer — APPROVED** [review]

**Pull request merged**